Update RemapGenes.R

souramoo opened the pull request:
1. Fix bug in `mapGenes` (countMatrix does not exist)

2. Make `mapReadGTF` more versatile (ignores comment lines and parser regex is more robust), so now it works on the latest 10x cellranger `refdata-gex-mm10-2020-A/genes/genes.gtf`)

Previously, the lack of support for ignoring comments lead to the error
```
Error in `mutate_cols()`:                                                                                                                                                                                                         
! Problem with `mutate()` column `gene_id`.
ℹ `gene_id = gsub(regxPattern, "\\1", Attributes)`.
x object 'Attributes' not found
Caused by error in `is.factor()`:
! object 'Attributes' not found
Run `rlang::last_error()` to see where the error occurred.
```

The previous regex parsers led to issues in the data frame being set up properly (because one cannot guarantee the "gene_source" attribute will immediately follow the gene_name)

souramoo committed:
Update RemapGenes.R

Fix bug in mapGenes and make mapReadGTF more versatile (ignores comment lines and parser is more robust, now works on latest cellranger refdata-gex-mm10-2020-A/genes/genes.gtf)

souramoo commented:
Thanks for the great project and paper! I was working on some scRNAseq analysis with some neurons, and we used the latest cellranger transcriptome reference to map onto. I have been playing around with projecting our data onto a reference (and yours seemed the most comprehensive drawing from multiple datasets!) to identify what spinal cord cells we can identify in our different conditions using seurat's `TransferData` and your reference.rds file.

I was starting to work on identifying the different neuronal subtypes present in each using your neural network. I ran into the bug at https://github.com/ArielLevineLabNINDS/SeqSeek_Classify_Full_Pipeline/issues/1 where the neural network was expecting a particular set of genes, and so I ran into the above bugs while tidying up my data to work while following the steps at https://colab.research.google.com/github/ArielLevineLabNINDS/SeqSeek-Classify-NN/blob/master/example_use_NN.ipynb ("so now you want to...".

danielruss commented:
@souramoo Thank you for the update. You are absolutely right that countMatrix should be count_matrix (switching from R/python/javascript/java makes it hard to remember conventions).   I downloaded the latest assembly from ebi and you are right; the regex is too specific.  However,  I have a question.  You chose to break it into two separate regex calls.  Any reason in particular?  I like the one-liner because it REQUIRES both a gene id AND a gene name on the same line.  

What do you think?

danielruss commented:
ok,  After more testing I see why you went with two separate regex patterns.  However, your code does not work for me.  For some reason the \\d  is not being recognized.  When I use  `gene_id \"(ENSMUSG[0-9\\.]+)\";.+` it appears to work fine.  I propose that you change the \\d to [0-9]. They should be the same. can you check if this change breaks anything on your side.

How I tested:
download and gunzip 
http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/gencode.vM23.primary_assembly.annotation.gtf.gz

```
x <- read_tsv("~/Downloads/gencode.vM23.primary_assembly.annotation.gtf",
             comment = "#",n_max = 100,col_names = FALSE)
x %>% mutate(x,
                   id1_found=grepl("gene_id \"(ENSMUSG\\d+)\";.+",X9),
                   id2_found=grepl("gene_id \"(ENSMUSG[0-9\\.]+)\";.+",X9),
                   name1_found=grepl(".+ gene_name \"([[:print:]]+?)\"; .+",X9),
                   gene_id  =gsub ("gene_id \"(ENSMUSG[0-9\\.]+)\";.+","\\1",X9 ),
                   gene_name  =gsub (".+ gene_name \"([[:print:]]+?)\"; .+","\\1",X9 ) ) %>%
  select(id1_found:gene_name)
```

Does this work for you?  I get  id1_found always false, but id2_found is true.

souramoo committed:
Update RemapGenes.R

Ensure regex matches gene ids that include a full stop, and combine two regex's into one.

souramoo commented:
Thanks for the fast review - and yes I can confirm I get the same results as you with the reference you have provided.

I have played around with the regex and combined them into one as desired.

The issue with `\d` is that it does exactly that and only looks for numerical sequences to match. However, the reference you provided contains full stops, e.g. `ENSMUSG00000102693.1` rather than `ENSMUSG00000102693` which the parser should also really be able to deal with, and so to use `\d` we would need to amend the regex to be `"gene_id \"(ENSMUSG[\\d\\.]+)\"`. I can confirm that all variants of these regexes work fine for my use case and agree the `[0-9]` syntax is probably a bit more explicit and have made the proposed changes to the pull request!

souramoo committed:
Update RemapGenes.R

Make diff easily comparable to main branch

souramoo committed:
Update RemapGenes.R

danielruss commented:
thank you @souramoo
